File: Stack/20. Valid Parentheses.py


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c == '(':
                stack.append(')')
            elif c == '{':
                stack.append('}')
            elif c == '[':
                stack.append(']')
            elif not stack or stack.pop() != c:
                return False

        return not stack


if __name__ == "__main__":
    s = "()[]"
    print(isValid(s))

File: Stack/20. Valid Parentheses.cpp
#include <bits/stdc++.h>

using std::cout, std::endl, std::string;


class Solution {
public:
    bool isValid(string s) {
        std::stack<char> parentheses;
        
        for (char& c: s)
        {
            if (c == '(')
                parentheses.push(')'); 
            else if (c == '[')
                parentheses.push(']');
            else if (c == '{')
                parentheses.push('}');
            else if (!parentheses.empty() && c == parentheses.top())
                parentheses.pop();
            else
                return false;
        }

        return parentheses.empty();
    }
};